### ISSUE TITLE
Including Rack context (url, method, ...) with Sentry appender?

### ISSUE THREAD

**Issue by tsauerwein** [rocketjob/semantic_logger#97]
### Environment
* Ruby Version: 2.3.0
* Semantic Logger Version: 4.2.0
* Application/framework names and versions: Sinatra

### Question

I am trying to use Semantic Logger together with Sentry. While the provided Sentry appender works fine (as in it sends messages to Sentry), it does not include the [Rack context](https://github.com/getsentry/raven-ruby/blob/9649fa5a50e5dcac9df108755990740af2c34822/lib/raven/integrations/rack.rb#L70-L81) (url, method, query string, ...) that is usually send with `Raven.capture_message` when the [Rack middleware](https://docs.sentry.io/clients/ruby/integrations/rack/) is activated. I figured out that this is because the appender runs in a separate thread, so that it does not have access to the Rack context values.

Digging a bit through the code, I found the [SemanticLogger.on_log](https://github.com/rocketjob/semantic_logger/blob/c94e811ec5d1137098eca733accb5c8468a20ded/lib/semantic_logger/semantic_logger.rb#L201-L227) hook. From the description, this sounds like a good place to access the context values or to directly call `Raven.capture_message` (instead of using the Sentry appender). But unfortunately, I am getting an error when trying to call `on_log`:

```
/usr/local/bundle/gems/semantic_logger-4.2.0/lib/semantic_logger/semantic_logger.rb:226:in `on_log': undefined method `on_log' for #<SemanticLogger::Appender::Async:0x00000000033fb328> (NoMethodError)
```
//edit: Here is a failing test for the above error: https://github.com/rocketjob/semantic_logger/compare/master...tsauerwein:on-log?expand=1

What is the recommended way to use Semantic Logger together with Sentry, so that the Rack context is included in the messages?

**Comment by reidmorrison** [comment]
Thank you for the fix for `on_log`. It would be great to get a PR with your changes for the Sentry appender to use `on_log` to add the Rack context, if you would like to share them?

**Comment by gingerlime** [comment]
@tsauerwein thanks a lot for posting this. I bumped into exactly the same problem, and wasn't sure what's happening or why. It seems obvious in hindsight that running in a separate thread would lose all this info.

In case it helps others, this is more or less what I ended up doing (mind you, this is using Rollbar, but the same principle applies to Sentry, Bugsnag etc I imagine)

```ruby
  SemanticLogger.on_log do |log|
    next unless log.try(:level) == :error
    err = RuntimeError.new(log.try(:message))
    err.set_backtrace(log.backtrace) if log.backtrace
    Rollbar.error(err, :log_extra => log.to_h)
  end
```

This would hook into `on_log` (instead of using an Appender) and then for every error being logged, would forward it to Rollbar.
